Add MCTP over USB Binding support

eric-ocasio-nxp opened the pull request:
Add USB device binding for MCTP. Created new USB device class for MCTP, libMCTP binding, and sample application.

henrikbrixandersen commented on doc/connectivity/usb/device_next/api/usbd_mctp.rst:
```diff
@@ -0,0 +1,11 @@
+.. _usbd_mctp:
+
+MCTP USB device API
```
Please detail what "MCTP" actually is here.

teburd commented on samples/modules/pmci/mctp/usb_endpoint/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+
+cmake_minimum_required(VERSION 3.20.0)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(mctp_endpoint)
```
project name could use something more descriptive, I get I'm guilty of probably starting this but perhaps mctp_usb_endpoint would be more descriptive

teburd commented on samples/modules/pmci/mctp/usb_endpoint/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+project(mctp_endpoint)
+
+include(${ZEPHYR_BASE}/samples/subsys/usb/common/common.cmake)
+FILE(GLOB app_sources src/*.c)
```
please don't glob here

teburd commented on samples/modules/pmci/mctp/usb_endpoint/README.rst:
```diff
@@ -0,0 +1,113 @@
+
+An example script to test this interface is provided below.
+
+.. code-block:: python
```
Honestly probably better to put this into scripts or into the sample directory itself as a python script with requirements.txt and all rather than pushing users to copy/paste, and figure out how to get the python deps installed.

Can always include whole code files into the docs like this with [literalinclude](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-literalinclude)

teburd commented:
Non-blocking comments obviously, but would be good to see improvements.

eric-ocasio-nxp commented on doc/connectivity/usb/device_next/usb_device.rst:
```diff
@@ -41,6 +41,8 @@ Samples
 
 * :zephyr:code-sample:`usb-hid-mouse`
 
+* :zephyr:code-sample:`mctp-usb-endpoint`
```
Do you suggest that I just update this file in a commit with the sample application, or do one 'super commit' with all of these feature/MCTP changes? I did it this way for human readability, but didn't fully understand the need for every commit to build.

eric-ocasio-nxp commented on subsys/usb/device_next/class/Kconfig.mctp:
```diff
@@ -0,0 +1,42 @@
+	help
+	  Address for the Bulk OUT endpoint.
+
+config USBD_MCTP_INTERFACE_SUBCLASS
+	int "Sub-class field of the interface descriptor"
+	default 0
+	help
+	  Sub-class field of the interface descriptor.
+
+config USBD_MCTP_INTERFACE_PROTOCOL
+	int "Protocol field for interface descriptor"
+	default 1
+	help
+	  Protocol field for interface descriptor.
```
https://www.usb.org/defined-class-codes, MCTP section.

tmon-nordic replied:
This should be a constant then, not a Kconfig option. Protocol is used to select MCTP version - how od you imagine Kconfig changing it magically?

If you want to allow version selection via Kconfig then addint MTP protocol `choice` is the approach and not controlling protocol field directly.

eric-ocasio-nxp replied:
I thought about this some more, and I think Kconfig is not the right spot for this. In the case there are multiple instances of a USB MCTP endpoint, the developer could want the flexibility of setting these on a per-instance basis, not globally. I will move these to the DT yaml file.

nashif commented on samples/modules/pmci/mctp/usb_endpoint/Kconfig:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2023 Nordic Semiconductor ASA
```
> I am confused, is PMCI/MCTP not part of the Zephyr but an external module?

mctp is not an external module, it is part of the main manifest and there are dependencies on it already.

jfischer-no replied:
> > I am confused, is PMCI/MCTP not part of the Zephyr but an external module?
> 
> mctp is not an external module, it is part of the main manifest and there are dependencies on it already.

Anyway, the implementation cannot be part of subsys/usb, and has to go to the modules.

dleach02 replied:
I believe that @jfischer-no is referring to zephyr\modules directory because that is where we put any dependency bindings between Zephyr and modules in the ecosystem. Am I correct in this interpretation Johann?

jfischer-no replied:
> The fact that MCTP is listed on usb.org doesn't matter then?

It does not mean anything. The specification is from DMTF association, and I already mentioned that, see https://github.com/zephyrproject-rtos/zephyr/pull/95347#discussion_r2356192394. You should know that, as you had to read it to write this code, or where does the code come from?

> I too am confused by the comment that this should go to modules, mctp is a subsystem in zephyr built on libmctp.

What is external dependency, and like a sample in this PR is placed under `sample/modules` the implementation should be in `modules/foobar`.

> yeah, this is confusing. The fact that libmctp is a module is irrelevant, this is implementing usb bindings, using both mctp and usb in zephyr, this has nothing to do with libmctp, the usb binding is a zephyr thing.

"usb binding" are useless without libmctp, "usb binding" is some trivial USB code that I will not accept in `subsys/usb` as long as there is no native libmctp implementation in Zephyr. And even then there is no need to place yet another backend to "something" in `subsys/usb`, a very good example is 6fddc44d590568666535097d25ef4ab34a68a0bf.

teburd replied:
@jfischer-no please, this is like saying I should not accept any code that depends on a hal, we've passed that rubicon long ago

To say it should go in subsys/pmci/mctp that's fine, its where the rest of the binding code is anyways. I get if you don't want to take maintainership of this code that's very understandable.

nashif replied:
> "usb binding" are useless without libmctp

right, but we have libmctp and it is not optional. 

Anyways, what is it exactly that you want?
- Is it about the location where the code should be hosted?
- Is it about this base class not being defined by USB-IF? Where should those go? What about bluetooth, not defined by USB-IF yet it is hosted in `subsys/usb/device/class`?
- Or is it about libmctp being a module? In this case, this is a library, mctp and bindings use this library, libmctp on its own does not do anything.

jfischer-no replied:
> @jfischer-no please, this is like saying I should not accept any code that depends on a hal, we've passed that rubicon long ago

We did not pass it anywhere. e.g. vendor-neutral driver code must not depend on any HAL.

> To say it should go in subsys/pmci/mctp that's fine, its where the rest of the binding code is anyways. I get if you don't want to take maintainership of this code that's very understandable and reasonable.

>right, but we have libmctp and it is not optional.

What in Zephyr depends on libmctp? Even MCTP samples are only located in samples/modules. MCTP is someone else's library, there are no dependencies in Zephyr, and I do not see any reason there should be any. The code in subsys/pmci/mctp, in contrast, depends on other Zephyr subsystems, whose maintainers have to care about it for no reason. And therefore I am confident that to make it optional is not even enough, it should be moved to external modules, see https://github.com/zephyrproject-rtos/zephyr/issues/91061#issuecomment-2957039567.

teburd replied:
> > @jfischer-no please, this is like saying I should not accept any code that depends on a hal, we've passed that rubicon long ago
> 
> We did not pass it anywhere. e.g. vendor-neutral driver code must not depend on any HAL.

Where is this written? Why do we have openthread then?

eric-ocasio-nxp replied:
@teburd ok, I'll explore this.

Regarding dependency in general, I want to clarify something since it may not be clear from this initial PR. The re-worked PR I have locally has cleaner lines of separation...

Besides the reference to the MCTP sample in usb_device.rst, the USB device class has no dependency on libmctp. Yes, libmctp uses it, but any other MCTP implementation can as well. You can remove libmctp and the device class stands on its own.

nashif replied:
> What in Zephyr depends on libmctp?

```
 git grep  libmctp.h subsys/ include/
include/zephyr/pmci/mctp/mctp_i2c_gpio_controller.h:#include <libmctp.h>
include/zephyr/pmci/mctp/mctp_i2c_gpio_target.h:#include <libmctp.h>
include/zephyr/pmci/mctp/mctp_uart.h:#include <libmctp.h>
subsys/pmci/mctp/mctp_memory.c:#include <libmctp.h>
```

nashif replied:
> Even MCTP samples are only located in samples/modules.

is this your issue? really?  https://github.com/zephyrproject-rtos/zephyr/pull/96995

eric-ocasio-nxp commented on subsys/usb/device_next/class/usbd_mctp.c:
```diff
@@ -0,0 +1,282 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(usbd_mctp, CONFIG_USBD_MCTP_LOG_LEVEL);
+
+#define DT_DRV_COMPAT zephyr_mctp_usb
```
I did it this way because there exists the possibility of having multiple MCTP buses/bindings. If the application requires more than one MCTP endpoint, a USB composite device could be used with more than one instance of USBD MCTP. Is there a way to accomplish this without DT?

eric-ocasio-nxp commented on subsys/usb/device_next/class/Kconfig.mctp:
```diff
@@ -0,0 +1,42 @@
+
+if USBD_MCTP_CLASS
+
+config USBD_MCTP_BULK_IN_EP_ADDR
+	hex "Address for the Bulk IN endpoint"
+	default 0x81
+	help
+	  Address for the Bulk IN endpoint.
+
+config USBD_MCTP_BULK_OUT_EP_ADDR
+	hex "Address for the Bulk OUT endpoint"
+	default 0x01
+	help
+	  Address for the Bulk OUT endpoint.
```
You're right, will remove. I had a fundamental misunderstanding about how endpoint numbering works.

eric-ocasio-nxp commented on subsys/usb/device_next/class/usbd_mctp.c:
```diff
@@ -0,0 +1,282 @@
+
+	struct udc_buf_info *bi = udc_get_buf_info(buf);
+
+	if (bi->ep == CONFIG_USBD_MCTP_BULK_OUT_EP_ADDR) {
+		if (ops->data_recv_cb != NULL) {
+			ops->data_recv_cb(dev, buf->data, buf->len, data->user_data);
+		}
+
+		k_work_submit(&data->out_work);
+	}
+
+	if (bi->ep == CONFIG_USBD_MCTP_BULK_IN_EP_ADDR) {
+		if (ops->tx_complete_cb != NULL) {
+			ops->tx_complete_cb(dev, data->user_data);
+		} else {
+			k_sem_give(&data->in_sem);
+		}
+	}
```
I'll look into adding a method to propagate errors, thanks.

eric-ocasio-nxp commented on subsys/usb/device_next/class/usbd_mctp.c:
```diff
@@ -0,0 +1,282 @@
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct usbd_mctp_data *data = dev->data;
+
+	k_sem_init(&data->in_sem, 1, 1);
```
Yes, that is how it's meant to work.

dleach02 commented:
@jfischer-no, @eric-ocasio-nxp, @teburd, @nashif 

Would it be helpful to move this to the arch WG call (@carlescufi) so we can collectively discuss what is needed to get this PR to move? It is unclear to me what the specific request is that will satisfy @jfischer-no change request and the back-and-forth in this PR doesn't seem to be productive?

dkalowsk commented:
@carlescufi @eric-ocasio-nxp @teburd it'd be a real shame if this PR doesn't make it for the 4.3 release.  Those of us building and shipping products utilizing MCTP benefit greatly from having the tighter integration with the mainline Zephyr releases.  Providing bindings like this ensure a clean path to enabling, reducing a lot of end user frustrations.

eric-ocasio-nxp committed:
pmci: mctp: samples: Add support for MCTP over USB.

Adds support for the USB device binding for MCTP. Binding created based on
the DMTF specification for USB Transport Binding. The overall design of
this binding follows the existing UART and I2C+GPIO bindings that already
exist in Zephyr.

Signed-off-by: Eric Ocasio <eric.ocasio@nxp.com>

nashif commented:
@jfischer-no  Please take another look.

kartben reviewed (approved):
+1 for the code sample - thx!

eric-ocasio-nxp commented:
I just updated the commit message since I goofed it up. Twice.

@teburd I think you'll need to re-approve.

kartben commented:
> weirdly I see the zephyrbot line requesting review from @kartben and @teburd. And it only shows +1 from myself and @nashif.
> 
> The bot is weird

yeah it depends where you look though, it seems... Github having a bad day maybe, who knows

<img width="550" height="370" alt="image" src="https://github.com/user-attachments/assets/1cf73030-37a7-4d4e-bd6b-6f6f621c7426" />

rruuaanng commented on include/zephyr/pmci/mctp/mctp_usb.h:
```diff
@@ -0,0 +1,90 @@
+ *
+ */
+
+#ifndef ZEPHYR_MCTP_USB_H_
```
```suggestion
#ifndef ZEPHYR_INCLUDE_PMCI_MCTP_USB_H_
```
